### PR TITLE
Make negated pattern position-insensitive

### LINKED ISSUE #29: Negated rule is position - sensitive
If the negated rule is the last that actually matches, then the result list will be empty, but otherwise not:
```
match(['ab'], ['a*', '!ab'])    //  --> []
match(['ab'], ['!ab', 'a*'])    //  --> ['ab']
match(['ab'], ['x', '!ab', 'a*'])    //  --> ['ab']
```
Because filtering is essentially a sets operation, one may expect that internal ordering of a set operand does not matter, but it does.

Not a big issue, but perhaps the readme should say something about this?

### PR DESCRIPTION
**Pull request by valango**
Maybe I was a bit hasty w this one, but it was blocking my long-dormant #25.

Fixes #29

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -76,8 +76,8 @@ module.exports = (inputs, patterns, options) => {
 		let matches = isFirstPatternNegated;
 
 		for (const pattern of patterns) {
-			if (pattern.test(input)) {
-				matches = !pattern.negated;
+			if (pattern.test(input) && !(matches = !pattern.negated)) {
```
Comment (sindresorhus) [inline index.js:79]: I don't like having an assignment in a condition. Move it out into a separate `if`.

File: readme.md [hunk 2]
```diff
@@ -118,7 +118,9 @@ Ensure you use this correctly. For example, files and directories should be matc
 
 Type: `string | string[]`
 
-Use `*` to match zero or more characters. A pattern starting with `!` will be negated.
+Use `*` to match zero or more characters.
+
+A leading `!` negates the pattern - any matching inputs will be excluded from the results.
```
Comment (sindresorhus) [inline readme.md:123]: You need to update index.d.ts too.

File: test.js [hunk 3]
```diff
@@ -5,6 +5,7 @@ test('matcher()', t => {
 	t.deepEqual(matcher(['foo', 'bar'], ['foo']), ['foo']);
 	t.deepEqual(matcher(['foo', 'bar'], ['bar']), ['bar']);
 	t.deepEqual(matcher(['foo', 'bar'], ['fo*', 'ba*', '!bar']), ['foo']);
+	t.deepEqual(matcher(['foo', 'bar'], ['fo*', '!bar', 'ba*']), ['foo']);
```
Comment (sindresorhus) [inline test.js:8]: This change will need a lot more assertions.

**Comment by sindresorhus** [comment]
You also need to update `.isMatch`.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
> String or array of string patterns.

This part feels moot in the TS definition (but not the readme) as it's already made clear by the types.

Otherwise, it looks ok.

**Comment by valango** [comment]
Plz check & let me know. I did not bump the version.

File: index.js [hunk 4]
```diff
@@ -65,30 +65,39 @@ module.exports = (inputs, patterns, options) => {
...
 	}
 
 	return result;
 };
 
+module.exports.isMatch = (inputs, patterns, options) => {
+	const matching = module.exports(inputs, patterns, options);
+
+	return matching.length > 0;
+};
```
Comment (sindresorhus) [inline index.js:99]: What's happening here? There's already an exported `.isMatch`.

File: index.js [hunk 5]
```diff
@@ -65,30 +65,39 @@ module.exports = (inputs, patterns, options) => {
...
 				matches = !pattern.negated;
+
+				if (!matches) {
+					break;
+				}
 			}
 		}
 
-		if (matches) {
+		if (matches || (matches === undefined && !patterns.some(p => !p.negated))) {
```
Comment (sindresorhus) [inline index.js:87]:
```suggestion
		if (matches || (matches === undefined && !patterns.some(pattern => !pattern.negated))) {
```

File: readme.md [hunk 6]
```diff
@@ -118,7 +118,12 @@ Ensure you use this correctly. For example, files and directories should be matc
 
 Type: `string | string[]`
 
-Use `*` to match zero or more characters. A pattern starting with `!` will be negated.
+Use `*` to match zero or more characters.
+
+A leading `!` negates the pattern.
+
+An input string will be omitted, if it does not match any non-negated patterns present,
+or if it matches a negated pattern, or if no pattern is present.
```
Comment (sindresorhus) [inline readme.md:126]: Don't hard-wrap.

File: readme.md [hunk 7]
```diff
@@ -118,7 +118,12 @@ Ensure you use this correctly. For example, files and directories should be matc
 
 Type: `string | string[]`
 
-Use `*` to match zero or more characters. A pattern starting with `!` will be negated.
+Use `*` to match zero or more characters.
+
+A leading `!` negates the pattern.
```
Comment (sindresorhus) [inline readme.md:123]: Readme and index.d.ts should be in sync documentation-wise.

File: test.js [hunk 8]
```diff
@@ -37,6 +41,10 @@ test('matcher.isMatch()', t => {
 	t.true(matcher.isMatch('moo', ['*oo', '!f*']));
 	t.true(matcher.isMatch('UNICORN', ['!*oo', 'UNI*'], {caseSensitive: true}));
```
Comment (sindresorhus) [inline test.js:43]: Move these to a separate test with a descriptive title of what they're testing. Same with the other assertions.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: readme.md (hunk 7 above)
Comment (valango) [reply readme.md:123]: "This part feels moot in the TS definition (but not the readme) as it's already made clear by the types."

Did I get you wrong or have you changed your mind?

File: test.js (hunk 8 above)
Comment (valango) [reply test.js:43]: Did so. New tests named "... negated patern placement".

File: readme.md (hunk 6 above)
Comment (valango) [reply readme.md:126]: Corrected.

File: index.js (hunk 5 above)
Comment (valango) [reply index.js:87]: Ok.

File: index.js (hunk 4 above)
Comment (valango) [reply index.js:99]: Fixed. [blush]

File: readme.md (hunk 7 above)
Comment (sindresorhus) [reply readme.md:123]: Ok, let me rephrase: index.d.ts and readme should be in sync, except for cases where the readme just points out the types as that's already clear by the types.

However, that's not what I'm pointing out here. You have removed some stuff from index.d.ts that is preserved here and not updated other parts, like https://github.com/sindresorhus/matcher/pull/30/files#diff-7aa4473ede4abd9ec099e87fec67fd57afafaf39e05d493ab4533acc38547eb8R83

File: index.js [hunk 9]
```diff
@@ -90,18 +93,7 @@ module.exports = (inputs, patterns, options) => {
...
-	}
-
-	return inputs.some(input => {
-		return patterns.every(pattern => {
-			const regexp = makeRegexp(pattern, options);
-			const matches = regexp.test(input);
-			return regexp.negated ? !matches : matches;
-		});
-	});
+	return matching.length > 0;
```
Comment (sindresorhus) [inline index.js:98]: The problem with this new logic here is that it's inefficient. The old way, it short-circuited when it found a match. Now it will do *everything* and then check for a match.

Comment (valango) [reply index.js:104]: Good point. I modified the code. Now isMatch seems to be bit quicker compared to the original implementation, probably because not constructing Regxp repeatedly.

File: index.js [hunk 10]
```diff
@@ -57,51 +57,49 @@ function makeRegexp(pattern, options) {
...
-module.exports.isMatch = (inputs, patterns, options) => {
-	inputs = sanitizeArray(inputs, 'inputs');
-	patterns = sanitizeArray(patterns, 'patterns');
+module.exports = (inputs, patterns, options) => matcher(inputs, patterns, options, false);
 
-	if (patterns.length === 0) {
-		return false;
-	}
+module.exports.isMatch = (inputs, patterns, options) => {
+	const matching = module.exports(inputs, patterns, options, true);
```
Comment (sindresorhus) [inline index.js:102]: This doesn't change anything as you're calling the incorrect method.

Comment (valango) [reply index.js]: Omg, made error on merge. Fixed now.

**Comment by sindresorhus** [comment]
You still need to do https://github.com/sindresorhus/matcher/pull/30#discussion_r624691927.

File: readme.md (hunk 7 above)
Comment (valango) [reply readme.md:123]: Ok, I restored the missing stuff and made the contents as coherent as I could via ee30306.
I also found: 
   - 'inputs' and 'patterns' were randomly used in both plural and singular form. Fixed that.
   - '@param options' is missing in index.d.ts - I left it for you to decide.

Comment (valango) [reply readme.md:123]: Ok, I restored the missing stuff and made the contents as coherent as coherent as I could via ee30306.
Also found: 
   - 'inputs' and 'patterns' were randomly used in both plural and singular form. Fixed that.
   - '@param options' is missing in index.d.ts - I left it for you to decide.
